display quoted notes #944

bryanmontz opened the pull request:
## Issues covered
#944 

## Description
This PR changes the "Link to note" rendering of links to notes in rendered note cards to show the note itself.

## How to test
1. Navigate to any note that refers to another note.
2. View the quoted note, tap on it to load.

## Screenshots/Video
| Before | After |
|--------|--------|
|![old](https://github.com/user-attachments/assets/601d94e6-9cb3-4ccd-a394-8792df144d25)|![new](https://github.com/user-attachments/assets/9e78ae5c-2a68-42ad-a78f-5208bf591c15)|

martindsq commented on Nos/Extensions/URL+Extensions.swift:
```diff
@@ -43,4 +43,8 @@ extension URL {
         }
     }
+    
+    var isValidNoteLink: Bool {
```
The computed var is almost self-explanatory, but I think there is still a benefit if we document it anyway so that new developers can understand what they are. Maybe a documentation entry explaining how note links look like and that they are generated by NoteParser would do.

bryanmontz replied:
Yeah, good point. However, I ended up not using this in the end, so I'll just remove it for now.

martindsq commented on Nos/Models/CoreData/Event+CoreDataClass.swift:
```diff
@@ -557,6 +558,9 @@ public class Event: NosManagedObject, VerifiableEvent {
         } else {
             let event = Event(context: context)
             event.identifier = id
+            context.perform {
```
This looks like something that could have side-effects in other parts of the app. Mind if I ping @mplorentz so he takes a look at it?

bryanmontz replied:
I moved it to avoid side-effects from other callers of this function.

mplorentz replied:
This does give me pause, because this function should only ever be called in a `context.perform` block already (but there's no way for you to know that @bryanmontz, it's just our convention) and I think I have seen nested context.perform blocks deadlock the app before. But this code has been on staging for some days and I haven't heard any reports of deadlock, so maybe it's fine? I guess I would rather it was reverted if you have the time. 

In general all the instance functions on NSManagedObjects are context agnostic and it's up to the caller to ensure that they are executed on the thread that Core Data wants. In the case of the view context it wants the main thread, and for all other contexts it wants to be in the context.perform block.

martindsq commented on Nos/Models/CoreData/Event+CoreDataClass.swift:
```diff
@@ -818,6 +820,13 @@ public class Event: NosManagedObject, VerifiableEvent {
+    @MainActor private(set) var quotedNoteID: RawEventID? {
+        didSet {
+            if quotedNoteID != nil {
+                Task { await loadFirstQuotedNote() }
```
Given that loadFirstQuotedNote() makes requests to relays, can you refactor this so setting quotedNoteID doesn't have side effects like this? It makes it harder to track what's happening in the app. I'm fine if that means having Task { await loadFirstQuotedNote() } in a couple of places. I think it would make the code easier to understand.

bryanmontz replied:
Moved it just to one spot.

martindsq commented on Nos/Models/CoreData/Event+CoreDataClass.swift:
```diff
@@ -826,18 +835,31 @@ public class Event: NosManagedObject, VerifiableEvent {
+            
+            await group.waitForAll()
         }
+        loadingViewData = false
```
This line of code was only called in the else condition before, I just want to double check having it here is desirable.

bryanmontz replied:
I believe that it is desirable. Otherwise, `loadingViewData` will never get set back to `false`, which could be misleading both to UI that depends on it as well as to devs trying to debug issues in this area. Would be great to get confirmation from @mplorentz if possible though.

martindsq reviewed (dismissed):
I'm excited to have this on main soon. Quoted notes are great fine for me (I'm not a big fan of the design though, I hope we polish it a little bit afterwards). I'm leaving some comments. I have testing it even more pending.

bryanmontz commented on NosTests/Models/NoteParserTests.swift:
```diff
@@ -233,20 +236,18 @@ final class NoteParserTests: CoreDataTestCase {
 
-        let expectedContent = "check this 🔗 Link to note. Bye!"
-        let (attributedContent, _) = sut.parse(
+        let expectedContent = "check this . Bye!"
```
I wasn't sure what to do with this. It's probably a Product question for Linda what to do when a note link is inline with other content as opposed to just at the end of a note.

joshuatbrown replied:
Yeah, I think it's fine to just leave like this for now, and maybe ask Linda's thoughts.

joshuatbrown commented on Nos/Models/NoteParser.swift:
```diff
@@ -15,34 +21,32 @@ struct NoteParser {
+    /// Parses the content and tags stored in a note and returns components that can be used
     /// to display the note in the UI.
-    func parse(content: String, tags: [[String]], context: NSManagedObjectContext) -> (AttributedString, [URL]) {
+    func parse(content: String, tags: [[String]], context: NSManagedObjectContext) -> NoteDisplayComponents {
```
I think it's time to rename this or the other `parse(content:tags:context:)` -- it's too confusing and annoying to have to specify the return type when we call either of them. I know you didn't name it, but would you mind fixing it? I'd suggest `components(...)` like Apple's Calendar API.

```suggestion
    func components(from content: String, tags: [[String]], context: NSManagedObjectContext) -> NoteDisplayComponents {
```

joshuatbrown reviewed (commented):
A few thoughts for now -- I didn't have time for a full review, but what I saw is looking great!

bryanmontz commented on Nos/Models/URLParser.swift:
```diff
@@ -3,7 +3,7 @@ import Logger
 /// Parses unformatted urls in a string and replace them with markdown links
 struct URLParser {
-    /// Returns an array with all unformated urls
+    /// Returns an array with all unformatted urls
```
I have a small, typos-only PR coming. 😉

joshuatbrown commented on Nos/Models/NoteParser.swift:
```diff
@@ -6,6 +6,12 @@ import RegexBuilder
+    struct NoteDisplayComponents {
+        let attributedContent: AttributedString
+        let contentLinks: [URL]
+        let quotedNoteID: RawEventID?
```
Yeah, I think you're right. I just wanted to think it through. I'm good with keeping it as a struct.

joshuatbrown commented on Nos/Models/NoteParser.swift:
```diff
@@ -6,6 +6,16 @@ import RegexBuilder
 /// This struct encapsulates the algorithms that parse notes and the mentions inside the note.
 struct NoteParser {
 
+    /// This struct encapsulates parsed components of a note that can be used to display the note in the UI.
```
Nitpick:
```suggestion
    /// Encapsulates parsed components of a note that can be used to display the note in the UI.
```
I always try to avoid using the type in the doc comments. I usually start by writing it exactly how you did here, then I delete the part where I said "This class" or "This struct" or whatever. 😀

bryanmontz replied:
Agree. I did that for consistency with the declaration of NoteParser right above it, but it's a different kind of thing. NoteParser is being used as a namespace here, and I think a protocol extension or caseless enum would be better for that.

joshuatbrown commented on Nos/Models/NoteParser.swift:
```diff
@@ -110,12 +118,19 @@ struct NoteParser {
+    ///
+    /// - Parameters:
+    ///   - content: The note content in which to replace entities.
+    ///   - capturesFirstNote: If true, this function will extract the first quoted note id, if it exists.
```
We've been adding default values to doc comments like this:

```suggestion
    ///   - capturesFirstNote: If true, this function will extract the first quoted note id, if it exists. Defaults to `false`.
```

joshuatbrown commented on Nos/Views/Modifiers/View+StyledBorder.swift:
```diff
@@ -0,0 +1,34 @@
+                    .strokeBorder(
+                        LinearGradient(
+                            gradient: Gradient(colors: [
+                                Color("styled-border-gradient-top"),
+                                Color("styled-border-gradient-bottom")
```
We use SwiftGen to generate color constants from the assets; you should be able to do something like this (but check my casing and spelling):

```suggestion
                                .styledBorderGradientTop,
                                .styledBorderGradientBottom
```

joshuatbrown commented on Nos/Views/Modifiers/View+StyledBorder.swift:
```diff
@@ -0,0 +1,34 @@
+                    )
+            )
+            .shadow(
+                color: Color(white: 0, opacity: 0.15),
```
Now that we have #1375 done, we'll probably want to create a new color constant for this. Maybe this would be named something like "styled-border-shadow" in the Colors.xcassets, then you should be able to use it here like this:

```suggestion
                color: .styledBorderShadow,
```

joshuatbrown commented on Nos/Views/Modifiers/View+StyledBorder.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+}
+
+struct StyledBorder: ViewModifier {
```
Would you mind adding a doc comment for this? Thanks!

joshuatbrown reviewed (commented):
This is looking sooooo great! I love how much better my feed looks with this! A few suggestions around documentation and SwiftGen, but otherwise this is awesome!

martindsq commented:
Thanks for addressing my points @bryanmontz I think we are now closer to finish this ticket.

bryanmontz commented:
@joshuatbrown and @martindsq I've addressed all the feedback. Anything else I can do to get your approval?

I know we'll need to merge `main` but I thought we could wait to do that until it's approved to avoid multiple merges.

joshuatbrown reviewed (commented):
Just one last question. This is looking awesome -- I love it!

bryanmontz commented on Nos/Views/NoteCard.swift:
```diff
@@ -8,25 +8,28 @@ import Dependencies
 struct NoteCard: View {
 
-    var note: Event
+    @ObservedObject var note: Event
```
It doesn't work in all cases with `let`.

joshuatbrown reviewed (approved):
Excellent work!

joshuatbrown reviewed (approved):
Latest updates look great!
